### ISSUE TITLE
how to populate bind options from data source (not manually)?

### ISSUE THREAD

**Issue by mattijn** [vega/vega#778]
I've modified the stocks example with a drop-down box to select a company, like this:
https://vega.github.io/new-editor/#/gist/vega/mattijn/a25b9ba41a775cf128d1472cd3e44253 (btw, great update to the new editor!)

But now I've to declare the `bind > options` manually (`"GOOG", "MSFT", etc`) I was hoping I could make use of a [Collect Transform](https://vega.github.io/vega/docs/transforms/collect/) to get all (unique) values from a data source, which I can use in the bind signal. But this seems not be supported?

**Comment by jheer** [comment]
I agree that would be useful, but it currently is not supported. I'll mark this as a feature request, but I don't know if/when we will get to it. So your best option may be to write your own JS code that generates the options list.

**Comment by Squaredj** [comment]
This would be really nice to have: for example to fill a "bind select" with items straight from the data itself. Any changes in the current Vega-release that would make this possible yet?

**Comment by AlexAbes** [comment]
@mattijn This is a great feature request. Did you end up writing any JS code that would generate the list of unique values for the options list as jheer suggested? Trying to figure this out myself while new to Vega, so any help you can give is appreciated.

**Comment by jheer** [comment]
@AlexAbes The following JS function will generate an array of unique values of the `property` field for a given `data` array. You can use the return value as the `options` within a signal binding:

```js
function uniqueValues(data, property) {
  var list = [];
  data.reduce(function(lookup, object) {
    var value = object[property];
    if (!lookup[value]) {
      list.push(value);
      lookup[value] = 1;
    }
    return lookup;
  }, {});
  return list; // you could optionally sort this list as well, if desired
}

**Comment by bphenriques** [comment]
This would also be useful in my use case, however since I am using Vega through Kibana, I can't generate the values on the fly. Therefore, binding to a specific data-field would be very useful.

**Comment by aidan-plenert-macdonald** [comment]
I'm interested in developing this feature. Before I get too deep, I was hoping to get some feed back.

For an example like [this](https://vega.github.io/vega/examples/regression/), I was thinking of just adding the syntax like,

```
     {
       "name": "groupby", "value": "none",
-      "bind": {"input": "select", "options": ["none", "genre"]}
+      "bind": {
+        "input": "select",
+        "options": {"data": "movies", "field": "Rotten_Tomatoes_Rating"}
+      }
     }
```

I figure that will fit nicely into the syntax. Any opinions or guidance?

**Comment by aidan-plenert-macdonald** [comment]
Just digging around the code, but it looks like the binds get initialized [here](https://github.com/vega/vega/blob/fd68b94aaa3626b0494da27d8201dc50b440a625/packages/vega-view/src/initialize.js#L39), and I need to make the change either [where the options are read in](https://github.com/vega/vega/blob/d82a2989bf2e32b4c02839c3aba8acf85b047e71/packages/vega-view/src/bind.js#L135) or where the [call to prepare element](https://github.com/vega/vega/blob/d82a2989bf2e32b4c02839c3aba8acf85b047e71/packages/vega-view/src/bind.js#L83) is.

It's not clear to me how to get at the actual data when the view is constructed. Is there a good example of how to read the data from within something like the view?

**Comment by aidan-plenert-macdonald** [comment]
vega-parser [constructs the Spec Scope with bindings](https://github.com/vega/vega/blob/cef824d53cf064d65c39aadac151bcd324b5bdd5/packages/vega-parser/src/Scope.js#L371) that get used in [View initialization](https://github.com/vega/vega/blob/cef824d53cf064d65c39aadac151bcd324b5bdd5/packages/vega-view/src/View.js#L65).

My test case dumps,

```
(index):119 TypeError: param.options.forEach is not a function
    at select (vega.js:31121)
    at generate (vega.js:31069)
    at bind$1 (vega.js:31034)
    at vega.js:31293
    at Array.forEach (<anonymous>)
    at View.initialize$1 [as initialize] (vega.js:31292)
    at new View (vega.js:34768)
```

The bind elements are constructed [here](https://github.com/vega/vega/blob/fd68b94aaa3626b0494da27d8201dc50b440a625/packages/vega-view/src/initialize.js#L39), which leads us to the [`forEach` call we see in the stack trace](https://github.com/vega/vega/blob/cef824d53cf064d65c39aadac151bcd324b5bdd5/packages/vega-view/src/bind.js#L135)

Seems like I could use [view.data](https://github.com/vega/vega/blob/cef824d53cf064d65c39aadac151bcd324b5bdd5/packages/vega-view/src/data.js#L12) ([see also](https://github.com/vega/vega/blob/cef824d53cf064d65c39aadac151bcd324b5bdd5/packages/vega-view/src/View.js#L344)) to get at the data.

So seems like I just need to stick something like below right [here](https://github.com/vega/vega/blob/cef824d53cf064d65c39aadac151bcd324b5bdd5/packages/vega-view/src/bind.js#L47),

```
if (param.options && !Array.isArray(param.options)) {
  if (param.options.field) {
    param.options = view.data(param.options.name).map(x => x[param.options.field]);
  } else {
    param.options = view.data(param.options.name);
  }
}
```

I'll test it out now.

**Comment by aidan-plenert-macdonald** [comment]
Not finished, but here is the current PR draft https://github.com/vega/vega/pull/1891

**Comment by mirjamle** [comment]
@jheer I am trying to implement your suggestion from 12 june 2018 to populate a select binding with unique values from my data:

> The following JS function will generate an array of unique values of the property field for a given data array. You can use the return value as the options within a signal binding.

I figured I can add the uniqueValues function through vega.expressionFunction and then use vega-embed to generate the visualization, like this:

```
vega.expressionFunction("uniqueValues", function uniqueValues(data, property) {
  var list = []; 
  data.reduce(function(lookup, object) {
    var value = object[property]; 
    if (!lookup[value]){
      list.push(value);
      lookup[value] = 1;
   } 
   return lookup;
 }, {}); 
return list;
})

const spec = "data/linegraph.vg.json";
vegaEmbed("#vis", spec)
          .then()
          .catch(console.warn);    
```
But I don't see how I can 

> use the return values as the options within a signal binding

I tried this:
```
"signals": [
    {
      "name": "signalname",
      "value": "",
      "bind": {
        "input": "select",
        "element": "#selectiondiv",
        "name": " ",
        "options": {"signal": "uniqueValues(mydata, 'propertyfield')"}
      }
    }
]
```

referencing the name of my data set and the property field I am interested in, but I then get 
TypeError: param.options.forEach is not a function

What is the proper way of doing this?

**Comment by aidan-plenert-macdonald** [comment]
@mirjamle Did you see my PR https://github.com/vega/vega/pull/1891?

**Comment by mirjamle** [comment]
@aidan-plenert-macdonald I noticed it, but I am new to Vega and not sure what it means...

**Comment by danmarshall** [comment]
FYI I just filed a related feature request #2152 for consideration, which suggests adding a backing `value` that corresponds to the displayed option.

**Comment by thegasta** [comment]
Any updates on this?

**Comment by scottclowe** [comment]
It would still be great to have this feature implemented.

**Comment by cafrank** [comment]
Still needed...
